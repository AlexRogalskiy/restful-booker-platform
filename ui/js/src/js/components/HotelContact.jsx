import React from 'react';
import validate from 'validate.js';

let rules = {
    "name": {
        presence : true,
        length: {
            minimum: 3,
            message: "must be at least 3 characters long"
        }
    },
    "email": {
        presence : true,
        email : true
    },
    "phone": {
        presence : true,
        numericality: {
            onlyInteger: true
        },
        length: {
            minimum: 11,
            message: "must be at least 11 digits long"
        }
    },
    "subject": {
        presence : true,
        length: {
            minimum: 20,
            message: "must be at least 20 characters long"
        }
    },
    "message" : {
        presence : true,
        length: {
            minimum: 20,
            message: "must be at least 20 characters long"
        }
    }
}

export default class HotelContact extends React.Component {

    constructor() {
        super();

        this.state = {
            name : "",
            email : "",
            phone : "",
            subject : "",
            message : "",
            submitted : false,
            errors : {
                name : [],
                email : [],
                phone : [],
                subject : [],
                message : [],
            }
        }

        this.updateState = this.updateState.bind(this);
        this.submitForm = this.submitForm.bind(this);
    }

    updateState(event){
        let currentState = this.state;

        currentState[event.target.id] = event.target.value;

        this.setState(currentState);
    }

    submitForm(){
        let vErrors = validate(this.state, rules);

        if(typeof vErrors === 'undefined'){
            this.setState({ submitted : true});
        } else {
            for (let prop in vErrors) {
                let currentState = this.state;
                currentState.errors[prop] = vErrors[prop];
                this.setState(currentState)
            }
        }
    }

    render(){
        let form = "";

        if(this.state.submitted){
            form = <div style={{height : "412px"}}>
                    <h2>Thanks for getting in touch {this.state.name}!</h2>
                    <p>We'll get back to you about</p>
                    <p style={{fontWeight : "bold"}}>{this.state.subject}</p>
                    <p>as soon as possible.</p>
                    </div>
        } else {
            form = <form>
                    <div className="input-group mb-3">
                        <div className="input-group-prepend">
                            <span className="input-group-text" id="basic-addon1"><span className="fa fa-id-card"></span></span>
                        </div>
                        <input type="text" className="form-control" placeholder="Name" aria-label="Name" id="name" aria-describedby="basic-addon1" onChange={this.updateState} />
                        {this.state.errors.name.map((value, index) => {
                            return <div key={"name" + index} className="invalid-feedback">{value}</div>
                        })}
                    </div>
                    <div className="input-group mb-3">
                        <div className="input-group-prepend">
                            <span className="input-group-text" id="basic-addon1"><span className="fa fa-envelope"></span></span>
                        </div>
                        <input type="text" className="form-control" placeholder="Email" aria-label="Email" id="email" aria-describedby="basic-addon1" onChange={this.updateState} />
                        {this.state.errors.email.map((value, index) => {
                            return <div key={"email" + index} className="invalid-feedback">{value}</div>
                        })}
                    </div>
                    <div className="input-group mb-3">
                        <div className="input-group-prepend">
                            <span className="input-group-text" id="basic-addon1"><span className="fa fa-phone"></span></span>
                        </div>
                        <input type="text" className="form-control" placeholder="Phone" aria-label="Phone" id="phone" aria-describedby="basic-addon1" onChange={this.updateState} />
                        {this.state.errors.phone.map((value, index) => {
                            return <div key={"phone" + index} className="invalid-feedback">{value}</div>
                        })}
                    </div>
                    <div className="input-group mb-3">
                        <div className="input-group-prepend">
                            <span className="input-group-text" id="basic-addon1"><span className="fa fa-envelope"></span></span>
                        </div>
                        <input type="text" className="form-control" placeholder="Subject" aria-label="Subject" id="subject" aria-describedby="basic-addon1" onChange={this.updateState} />
                        {this.state.errors.subject.map((value, index) => {
                            return <div key={"subject" + index} className="invalid-feedback">{value}</div>
                        })}
                    </div>
                    <div className="input-group">
                        <div className="input-group-prepend">
                            <span className="input-group-text">Message</span>
                        </div>
                        <textarea className="form-control" aria-label="Message" id="message" rows="5" onChange={this.updateState}></textarea>
                        {this.state.errors.message.map((value, index) => {
                            return <div key={"message" + index} className="invalid-feedback">{value}</div>
                        })}
                    </div>
                    <br />
                    <button type='button' className='btn btn-outline-primary float-right' id="submitContact" onClick={this.submitForm}>Submit</button>
                </form>
        }

        if(this.props.contact){
            return <div className='row contact'>
                        <div className='col-sm-1'></div>
                        <div className='col-sm-5'>
                            {form}
                        </div>
                        <div className='col-sm-5'>
                            <p><span className="fa fa-home"></span> {this.props.contact.name}</p>
                            <p><span></span> {this.props.contact.address}</p>
                            <p><span className="fa fa-phone"></span> {this.props.contact.phone}</p>
                            <p><span className="fa fa-envelope"></span> {this.props.contact.email}</p>
                        </div>
                        <div className='col-sm-1'></div>
                    </div>
        } else {
            return <p>ERROR: Missing contact details to render</p>
        }
    }

}