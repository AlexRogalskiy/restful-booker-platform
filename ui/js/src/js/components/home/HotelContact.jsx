import React from 'react';

export default class HotelContact extends React.Component {

    render(){
        if(this.props.contact){
            return <div>
                        <ul>
                            <li>{this.props.contact.name}</li>
                            <li>{this.props.contact.address}</li>
                            <li>{this.props.contact.phone}</li>
                            <li>{this.props.contact.email}</li>
                        </ul>
                    </div>
        } else {
            return <p>ERROR: Missing contact details to render</p>
        }
    }

}