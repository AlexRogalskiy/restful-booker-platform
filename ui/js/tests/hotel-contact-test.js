import React from 'react';
import HotelContact from '../src/js/components/HotelContact.jsx';

test('Contact form returns error messages', () => {
    const contactState = {
        name : "Mark",
        email : "mark@email.com",
        phone : "074123291234",
        subject : "I want cake",
        message : "This is my message for cake",
    }

    const hotelContactComponent = shallow(
        <HotelContact contact={contactState} />
    );

    hotelContactComponent.find('#submitContact').simulate('click');

    expect(hotelContactComponent).toMatchSnapshot();
});

test('Contact form can be submitted', () => {
    const contactState = {
        name: "Shady Meadows B&B",
        address: "The Old Farmhouse, Shady Street, Newfordburyshire, NE1 410S",
        phone: "0123456789",
        email: "fake@fakeemail.com"
    }

    const hotelContactComponent = shallow(
        <HotelContact contact={contactState} />
    );

    hotelContactComponent.setState({
        name : 'Mark',
        email : 'email@test.com',
        phone : '018392391183',
        subject : 'I want to book a room',
        message : 'And I want a bottle of wine with the booking',
    })
    hotelContactComponent.update();
    hotelContactComponent.find('#submitContact').simulate('click');

    expect(hotelContactComponent).toMatchSnapshot();
});