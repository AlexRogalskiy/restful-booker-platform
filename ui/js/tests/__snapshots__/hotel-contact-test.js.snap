// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact form can be submitted 1`] = `
<div
  className="row contact"
>
  <div
    className="col-sm-1"
  />
  <div
    className="col-sm-5"
  >
    <div
      style={
        Object {
          "height": "412px",
        }
      }
    >
      <h2>
        Thanks for getting in touch 
        Mark
        !
      </h2>
      <p>
        We'll get back to you about
      </p>
      <p
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        I want to book a room
      </p>
      <p>
        as soon as possible.
      </p>
    </div>
  </div>
  <div
    className="col-sm-5"
  >
    <p>
      <span
        className="fa fa-home"
      />
       
      Shady Meadows B&B
    </p>
    <p>
      <span />
       
      The Old Farmhouse, Shady Street, Newfordburyshire, NE1 410S
    </p>
    <p>
      <span
        className="fa fa-phone"
      />
       
      0123456789
    </p>
    <p>
      <span
        className="fa fa-envelope"
      />
       
      fake@fakeemail.com
    </p>
  </div>
  <div
    className="col-sm-1"
  />
</div>
`;

exports[`Contact form returns error messages 1`] = `
<div
  className="row contact"
>
  <div
    className="col-sm-1"
  />
  <div
    className="col-sm-5"
  >
    <form>
      <div
        className="input-group mb-3"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
            id="basic-addon1"
          >
            <span
              className="fa fa-id-card"
            />
          </span>
        </div>
        <input
          aria-describedby="basic-addon1"
          aria-label="Name"
          className="form-control"
          id="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
        />
        <div
          className="invalid-feedback"
          key="name0"
        >
          Name must be at least 3 characters long
        </div>
      </div>
      <div
        className="input-group mb-3"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
            id="basic-addon1"
          >
            <span
              className="fa fa-envelope"
            />
          </span>
        </div>
        <input
          aria-describedby="basic-addon1"
          aria-label="Email"
          className="form-control"
          id="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
        />
        <div
          className="invalid-feedback"
          key="email0"
        >
          Email is not a valid email
        </div>
      </div>
      <div
        className="input-group mb-3"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
            id="basic-addon1"
          >
            <span
              className="fa fa-phone"
            />
          </span>
        </div>
        <input
          aria-describedby="basic-addon1"
          aria-label="Phone"
          className="form-control"
          id="phone"
          onChange={[Function]}
          placeholder="Phone"
          type="text"
        />
        <div
          className="invalid-feedback"
          key="phone0"
        >
          Phone is not a number
        </div>
        <div
          className="invalid-feedback"
          key="phone1"
        >
          Phone must be at least 11 digits long
        </div>
      </div>
      <div
        className="input-group mb-3"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
            id="basic-addon1"
          >
            <span
              className="fa fa-envelope"
            />
          </span>
        </div>
        <input
          aria-describedby="basic-addon1"
          aria-label="Subject"
          className="form-control"
          id="subject"
          onChange={[Function]}
          placeholder="Subject"
          type="text"
        />
        <div
          className="invalid-feedback"
          key="subject0"
        >
          Subject must be at least 20 characters long
        </div>
      </div>
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
          >
            Message
          </span>
        </div>
        <textarea
          aria-label="Message"
          className="form-control"
          id="message"
          onChange={[Function]}
          rows="5"
        />
        <div
          className="invalid-feedback"
          key="message0"
        >
          Message must be at least 20 characters long
        </div>
      </div>
      <br />
      <button
        className="btn btn-outline-secondary float-right"
        id="submitContact"
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
  <div
    className="col-sm-5"
  >
    <p>
      <span
        className="fa fa-home"
      />
       
      Mark
    </p>
    <p>
      <span />
       
    </p>
    <p>
      <span
        className="fa fa-phone"
      />
       
      074123291234
    </p>
    <p>
      <span
        className="fa fa-envelope"
      />
       
      mark@email.com
    </p>
  </div>
  <div
    className="col-sm-1"
  />
</div>
`;
